File: Scala/Learn/src/main/scala/libs/Heap.scala
package libs.Heap
import scala.util.control.Breaks._

class Heap[T <: Comparable[T]] (var pool : List[T], 
            val compare : (T, T) => Boolean) {
    def size(): Int = {
        return this.pool.size;
    }

    def sift_up(index : Int) : Unit = {
        var i = index
        while ((i >> 1) > 0) {
            val fa = i >> 1;
            if (this.compare(this.pool(i), this.pool(fa))) {
                val temp = this.pool(fa)
                this.pool.updated(fa, this.pool(i))
                this.pool.updated(i, temp)
                i = fa

            } else {
                break
            }
        }
    }

    def sift_down(index : Int) : Unit = {
        var i = index
        while ((i << 1) < this.size()) {
            var cur = i
            val lson = (i << 1) + 1
            val rson = (i << 1) + 2
            if (this.compare(this.pool(lson), this.pool(cur)))
                cur = lson
            
            if (rson < this.size() && this.compare(this.pool(rson), this.pool(cur)))
                cur = rson
            
            if (cur != i) {
                val tmp = pool(cur)
                pool.updated(cur, pool(index))
                pool.updated(i, tmp)
                i = cur
            } else {
                break
            }
        }
    }

    def empty() : Boolean = {
        this.size == 0
    }

    def insert(dat : T) = {
        this.pool = this.pool :+ dat
        sift_up(this.size() - 1)
    }

    def pop() : Option[T] = {
        if (this.size() == 0) {
            None
        } else {
            val result = this.pool(0)
            this.pool = this.pool.tail
            pool match {
                case xs :+ last => {
                    pool = last :: xs
                    sift_down(0)
                }
                case Nil => Nil
            }
            Some(result)
        }
    }
}

File: Rust/Experiments/src/libs/heap.rs
use std;
use std::mem;
use std::clone;

#[derive(Debug)]
pub struct Heap<T> {
    size: usize,
    pool: Vec<T>
}

impl<T : Ord + Clone> Heap<T> {
    pub fn get_size(&self) -> usize {
        self.pool.len()
    }

    fn sift_up(&mut self, mut index: usize) {
        while (index >> 1) >= 0 {
            if self.pool[index] < self.pool[index >> 1] {
                self.pool.swap(index, index >> 1);
                index >>= 1;
            } else {
                break;
            }
        }
    }

    fn sift_down(&mut self, mut index: usize) {
        while ((index << 1) + 1) < self.get_size() {
            let lson = (index << 1) + 1;
            let rson = (index << 1) + 2;
            let mut cur = index;
            if self.pool[cur] > self.pool[lson] {
                cur = lson;
            }
            if rson < self.get_size()
                && self.pool[cur] > self.pool[rson] {
                cur = rson;
            }
            if cur != index {
                self.pool.swap(cur, index);
                index = cur;
            } else {
                break;
            }
        }
    }

    pub fn insert(&mut self, datum: T) {
        self.size += 1;
        self.pool.push(datum);
        self.sift_up(self.size - 1);
    }

    pub fn pop(&mut self) -> Option<T> {
        if self.get_size() > 0 {
            let ret = Some(self.pool[0].clone());
            match self.pool.pop() {
                Some(datum) => {
                    self.size -= 1;
                    if self.get_size() > 0 {
                        self.pool[0] = datum;
                        self.sift_down(0);
                    }
                    ret
                }
                None => None
            }
        } else {
            None
        }
    }

    pub fn empty(&self) -> bool {
        self.pool.len() == 0
    }

    pub fn new() -> Heap<T> {
        let ret = Heap {size : 0, pool : std::vec::Vec::new()};
        ret
    }
}